Lab5/src/Parser.java

```diff
@@ -0,0 +1,291 @@
+                    List<String> rightNonTerminals = new ArrayList<>();
+                    String rightFoundTerminal = null;
+                    for (String item : production)
+                        if (this.grammar.getNonTerminals().contains(item))
+                            rightNonTerminals.add(item);
+                        else {
+                            rightFoundTerminal = item;
+                            break;
+                        }
+                    concatenatedSet.addAll(concatenationOfLengthOne(rightNonTerminals, rightFoundTerminal));
+                }
+                if (!concatenatedSet.equals(firstForNonterminals.get(nonTerminal))) {
+                    isDifferent = true;
+                }
+                newIteration.put(nonTerminal, concatenatedSet);
+            }
+            firstForNonterminals = newIteration;
+        }
+    }
+
+    private Set<String> concatenationOfLengthOne(List<String> nonTerminals, String terminal) {
+        if (nonTerminals.size() == 0)
+            return new HashSet<>();
+        if (nonTerminals.size() == 1) {
+            return firstForNonterminals.get(nonTerminals.get(0));
+        }
+        Set<String> concatenation = new HashSet<>();
+        var i = 0;
+
+        while (i < nonTerminals.size()) {
+            boolean isEpsilon = false;
+            for (String s : firstForNonterminals.get(nonTerminals.get(i)))
+                if (s.equals("epsilon"))
+                    isEpsilon = true;
+                else
+                    concatenation.add(s);
+            if (isEpsilon)
+                i++;
+            else
+                break;
+        }
+
+        var allNonTerminalsHaveEpsilon = true;
+        for (String nonTerminal : nonTerminals) {
+            if (!firstForNonterminals.get(nonTerminal).contains("epsilon")) {
+                allNonTerminalsHaveEpsilon = false;
+            }
+        }
+        if (allNonTerminalsHaveEpsilon) {
+            concatenation.add(Objects.requireNonNullElse(terminal, "epsilon"));
+        }
+
+        return concatenation;
+    }
+
+
+    public void follow() {
+        for (String nonterminal : grammar.getNonTerminals()) {
+            followForNonterminals.put(nonterminal, new HashSet<>());
+        }
+        followForNonterminals.get(grammar.getStartSymbol()).add("epsilon");
+        boolean repeat = true;
+        while (repeat) {
+            repeat = false;
+            for (String nonterminal : grammar.getNonTerminals()) {
+                Set<String> prev = new HashSet<>(followForNonterminals.get(nonterminal));
+                Map<String, List<List<String>>> productionsContainingNonterminal = getProductionsContainingNonterminal(nonterminal);
+                for (Map.Entry<String, List<List<String>>> production : productionsContainingNonterminal.entrySet()) {
+                    for (List<String> elem : production.getValue()) {
+                        if (elem.indexOf(nonterminal) == elem.size() - 1)
```

Comment:
In situations where there are several nested statements it's becoming very difficult to follow stuff around
Maybe breaking it in multiple variables with more suggestive names would be helpful

Reply:
You've merely embraced nested statements...

![abc](https://user-images.githubusercontent.com/72268293/212119440-094684a2-66c5-4625-9117-b0dd736a29fe.png)